### PR TITLE
Add force decoding in FasterTransformer

### PR DESCRIPTION
**Pull request by FrostML**
### PR types

New features

### PR changes

APIs

### Description

Add force decoding for FasterTransformer. 
* Beam Search
* T2T Beam Search
* Sampling

### CODE REVIEW

**Comment by ZeyuChen** [comment]
Can we close this PR? @FrostML

**Comment by FrostML** [comment]
> Can we close this PR? @FrostML

@guoshengCS is reviewing this PR. 
This PR provides force decoding support in FasterTransformer. Need to be merged. 

This PR may cause lots of conflicts when merging into FT upgrading work. So need to be merged ASAP before resolving conflicts become a challenge..

**Comment by ZeyuChen** [comment]
> > Can we close this PR? @FrostML
> 
> @guoshengCS is reviewing this PR. This PR provides force decoding support in FasterTransformer. Need to be merged.
> 
> This PR may cause lots of conflicts when merging into FT upgrading work. So need to be merged ASAP before resolving conflicts become a challenge..

Got it. Please go ahead.

File: paddlenlp/ops/patches/FasterTransformer/decoding_beamsearch.h [hunk 1]
```diff
@@ -669,17 +723,19 @@ class DecodingBeamsearch {
...
-      // TODO Find a better method to check the is_finished
-      int finish_size = keep_alive_beam_ == true ? m * 2 : m;
-      cudaMemcpy(h_finished_buf_,
-                 finished_buf_,
-                 sizeof(bool) * finish_size,
-                 cudaMemcpyDeviceToHost);
-      int sum = 0;
-      for (int i = 0; i < finish_size; i++) {
-        sum += (int)h_finished_buf_[i];
+      if (step > min_trg_len) {
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/decoding_beamsearch.h]: trg_word中不包含结束的话是否可以`if (step > max_trg_len)`

File: paddlenlp/ops/patches/FasterTransformer/decoding_beamsearch.h [hunk 2]
```diff
@@ -611,30 +633,62 @@ class DecodingBeamsearch {
...
+            cum_log_buf_,
+            keep_alive_beam_,
+            args_.batch_size_,
+            (keep_alive_beam_) ? args_.beam_width_ * 2 : args_.beam_width_,
+            max_trg_len,
+            step,
+            decoding_params.stream);
       }
 
 #ifndef NDEBUG
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/decoding_beamsearch.h:694]: `update_KV_cache_kernelLauncher`是否也可以只在`step > min_trg_len`是进行

Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/decoding_beamsearch.h:694]: 可以，不过需要修改前面 `kv_cache_id`，否则因为是使用的 two-way buffer，还是得需要拷贝。连带的，可能还需要再  patch 一下 `update_KV_cache_kernelLauncher `，这个函数目前在 `decoding_xxx.h` 里面未曾 patch，在 `transformer_xxx.h` 里面有对应的 patch，所以是否可以记一个 TODO，在升级、合并这份加速代码里面看看怎么做

File: paddlenlp/ops/patches/FasterTransformer/decoding_beamsearch.h (hunk 1 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/decoding_beamsearch.h]: Done. Thanks.

**Review by guoshengCS — APPROVED** [review]

**Pull request merged**